### PR TITLE
17778 & 17788 - Org Admin/Coordinator for affiliation invitation updates

### PR DESCRIPTION
**Pull request by hfekete**
*Issue #:17778 & 17788*
https://github.com/bcgov/entity/issues/17778
https://github.com/bcgov/entity/issues/17788

*Description of changes:*
- updating how emails are fetched for affiliation invitation of type REQUEST (going to all admins/coordinators)
- updating who can see affiliation invitations of type REQUEST (only admin/coordinators now)
- updating who can accept affiliation invitations of type REQUEST (only admin/coordinators now)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 1]
```diff
@@ -362,10 +363,39 @@ def delete_affiliation_invitation(invitation_id):
...
+                                           org_id: int,
+                                           user: UserService = None) -> List[AffiliationInvitationModel]:
+        """Filter out affiliation invitations of type REQUEST if current user is not staff or org admin/coordinator."""
+        if UserService.is_context_user_staff():
+            return affiliation_invitation_models
+
+        if org_id is None:
+            return []
+
+        current_user: UserService = user or UserService.find_by_jwt_token()
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:377]: Look at the `@user_context` decorator?

https://github.com/bcgov/sbc-auth/blob/main/auth-api/src/auth_api/utils/user_context.py

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 2]
```diff
@@ -362,10 +363,39 @@ def delete_affiliation_invitation(invitation_id):
...
+        is_org_admin_or_coordinator = UserService.is_user_admin_or_coordinator(user=current_user, org_id=org_id)
+        if is_org_admin_or_coordinator:
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:379]: if UserService.is_user_admin_or_coordinator(user=current_user, org_id=org_id):
 return affiliation_invitation_models

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 3]
```diff
@@ -362,10 +363,39 @@ def delete_affiliation_invitation(invitation_id):
         else:
             invitation.delete()
 
+    @staticmethod
+    def _filter_request_invites_role_based(affiliation_invitation_models: List[AffiliationInvitationModel],
+                                           org_id: int,
+                                           user: UserService = None) -> List[AffiliationInvitationModel]:
```
Comment (ochiu) [inline auth-api/src/auth_api/services/affiliation_invitation.py:369]: It doesn't look like there is an explicit use of the user here, and the comment indicates current user.

Can we remove this param?

File: auth-api/src/auth_api/services/user.py [hunk 4]
```diff
@@ -648,12 +648,15 @@ def is_context_user_staff(**kwargs):
...
-        if not user:
-            return False
-
+    def is_user_admin_or_coordinator(user, org_id: int) -> bool:
+        """Check if user(userservice wrapper) provided is admin or coordinator for the given org id."""
         current_user_membership: MembershipModel = \
             MembershipModel.find_membership_by_user_and_org(user_id=user.identifier, org_id=org_id)
 
-        return current_user_membership is not None and current_user_membership.status == Status.ACTIVE.value
+        if current_user_membership is None:
```
Comment (seeker25) [inline auth-api/src/auth_api/services/user.py:656]: You need to handle the STAFF scenario here. BCREG STAFF can view other org's dashboards. They might not necessarily have a membership to the org.

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 5]
```diff
@@ -362,10 +363,39 @@ def delete_affiliation_invitation(invitation_id):
...
+        return list(filter(
+            lambda affiliation_invitation: affiliation_invitation.type != AffiliationInvitationType.REQUEST.value,
+            affiliation_invitation_models))
+
     @staticmethod
     def search_invitations(search_filter: AffiliationInvitationSearch, mask_email=True):
         """Search affiliation invitations."""
-        invitation_models = AffiliationInvitationModel().filter_by(search_filter=search_filter)
+        try:
+            org_id = int(search_filter.to_org_id or search_filter.from_org_id)
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:391]: Not sure if this is better or not? 
```
org_id_str = search_filter.to_org_id or search_filter.from_org_id
org_id = int(org_id_str) if org_id_str  and org_id_str .isdecimal() else None
```

File: auth-api/src/auth_api/services/affiliation_invitation.py [hunk 6]
```diff
@@ -362,10 +363,39 @@ def delete_affiliation_invitation(invitation_id):
         else:
             invitation.delete()
 
+    @staticmethod
+    def _filter_request_invites_role_based(affiliation_invitation_models: List[AffiliationInvitationModel],
+                                           org_id: int,
+                                           user: UserService = None) -> List[AffiliationInvitationModel]:
+        """Filter out affiliation invitations of type REQUEST if current user is not staff or org admin/coordinator."""
+        if UserService.is_context_user_staff():
```
Comment (seeker25) [inline auth-api/src/auth_api/services/affiliation_invitation.py:371]: awesome

**Review by seeker25 — APPROVED** [review]

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 5 above)
Comment (hfekete) [reply auth-api/src/auth_api/services/affiliation_invitation.py:389]: Yeah, I was trying to decide before writing it. Cause I like how some things are handled for these kind of issues in JS. But I feel like it does not translate very well into the Python. So will leave it as is.

File: auth-api/src/auth_api/services/affiliation_invitation.py (hunk 3 above)
Comment (hfekete) [reply auth-api/src/auth_api/services/affiliation_invitation.py]: Yeah, was passing it previously from another method, but after bit of refactor its not being passed anymore. And agreed, if its not explicitly used, remove!
Thanks for catching it.

**Pull request merged**